### PR TITLE
feat: allow nested `requirements.txt` files

### LINKED ISSUE #3102: feature: allow nested `requirements.txt` files
### Feature request

I would like to be able to have a `requirements.txt` file which references other requirements files with `-r`.

### Motivation

I would like to be able to you `pip-compile-multi` to manage the dependencies of my models. Often there are different dependencies at serve and train time with some overlap. Therefore, I'd like to be able to define the common dependencies in one file, the serving specific ones in another file, and the training specific ones in a third file.

ie.

**model.txt**
```
scikit-learn
```

**train.txt**
```
-r model.txt
pandas
```

**serve.txt**
```
-r model.txt
pydantic
```

### Other

_No response_

### PR DESCRIPTION
**Pull request by judahrand**
## What does this PR address?

This is arguably a bug fix? Not sure... somewhere between a bug fix and a feature... I'd expected this to 'just work'.

Fixes #3102

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [x] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [x] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_internal/utils/__init__.py [hunk 1]
```diff
@@ -303,6 +304,24 @@ def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str:
     raise FileNotFoundError(f"file {filepath} not found")
 
 
+def find_nested_requirements_txts(
+    base_requirements_txt: Path,
+) -> t.Tuple[Path, t.Set[Path]]:
+    """Recursively search the provided requirements.txt file to find nested
+    requirements files specified with `-r`.
+    """
+    sub_requirements_txt_files: t.Set[Path] = set()
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/__init__.py:313]:
```suggestion
    sub_requirements_txt_files: set[Path] = set()
```

File: src/bentoml/_internal/utils/__init__.py [hunk 2]
```diff
@@ -303,6 +304,24 @@ def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str:
     raise FileNotFoundError(f"file {filepath} not found")
 
 
+def find_nested_requirements_txts(
+    base_requirements_txt: Path,
+) -> t.Tuple[Path, t.Set[Path]]:
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/__init__.py:309]:
```suggestion
) -> tuple[Path, set[Path]]:
```

**Review by aarnphm — COMMENTED** [review]
for this feature I feel that we should use a requirements parser https://github.com/sarugaku/requirementslib or https://github.com/nexB/pip-requirements-parser

File: src/bentoml/_internal/utils/__init__.py [hunk 3]
```diff
@@ -279,6 +279,7 @@ def copy_file_to_fs_folder(
     src_fs = fs.open_fs(dir_name)
     dst_filename = file_name if dst_filename is None else dst_filename
     dst_path = fs.path.join(dst_folder_path, dst_filename)
+    dst_fs.makedir(dst_folder_path, recreate=True)
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/__init__.py:282]: I don't think this is correct.

File: src/bentoml/_internal/bento/build_config.py [hunk 4]
```diff
@@ -544,15 +546,25 @@ def write_to_bento(self, bento_fs: FS, build_ctx: str) -> None:
...
                 dst_filename="requirements.txt",
             )
+            for txt in nested_txts:
+                dst_path = txt.relative_to(base_txt.parent)
+                copy_file_to_fs_folder(
+                    str(txt.absolute()),
+                    bento_fs,
+                    fs.path.combine(py_folder, str(dst_path.parent)),
+                    dst_filename=dst_path.name,
+                )
```
Comment (aarnphm) [inline src/bentoml/_internal/bento/build_config.py:567]: I propose that for requirements.txt alone, we flatten all requirements into one requirements.txt.

File: src/bentoml/_internal/utils/__init__.py (hunk 1 above)
Comment (judahrand) [reply src/bentoml/_internal/utils/__init__.py:313]: That's only valid syntax in Python 3.9+ isn't it?

File: src/bentoml/_internal/utils/__init__.py (hunk 2 above)
Comment (judahrand) [reply src/bentoml/_internal/utils/__init__.py:309]: That's only valid syntax in Python 3.9+ isn't it?

File: tests/e2e/bento_server_http/bentofile.yaml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 service: service:svc
 python:
-  requirements_txt: ./requirements.txt
+  requirements_txt: ./requirements/service.txt
```
[the remark this answers is not in the document]
Comment (judahrand) [reply tests/e2e/bento_server_http/bentofile.yaml:3]: Why?

**Comment by judahrand** [comment]
> for this feature I feel that we should use a requirements parser https://github.com/sarugaku/requirementslib or https://github.com/nexB/pip-requirements-parser

Why? Do we need to introduce an additional dependency for this fix/feature? Do we gain anything?

**Comment by judahrand** [comment]
Happy to be wrong on any of my questions but I'm not sure I understand your reasons for the proposed changes.

File: src/bentoml/_internal/utils/__init__.py (hunk 2 above)
Comment (sauyon) [reply src/bentoml/_internal/utils/__init__.py:309]: Type annotations aren't evaluated with `from __future__ import annotations`, so we use the new syntax for type annotations.

File: src/bentoml/_internal/bento/build_config.py [hunk 6]
```diff
@@ -544,15 +546,25 @@ def write_to_bento(self, bento_fs: FS, build_ctx: str) -> None:
...
             )
             copy_file_to_fs_folder(
-                requirements_txt_file,
+                str(base_txt.absolute()),
                 bento_fs,
                 py_folder,
                 dst_filename="requirements.txt",
             )
+            for txt in nested_txts:
+                dst_path = txt.relative_to(base_txt.parent)
```
Comment (sauyon) [inline src/bentoml/_internal/bento/build_config.py:561]: What does `dst_path` resolve to for `-r ../../requirements.txt`?

Comment (judahrand) [reply src/bentoml/_internal/bento/build_config.py:561]: It throws an exception. I should have noted somewhere that there is a requirement that nested files live at the same level or below the base file. To me this feels like a reasonable limitation.

File: src/bentoml/_internal/utils/__init__.py (hunk 1 above)
Comment (aarnphm) [reply src/bentoml/_internal/utils/__init__.py:313]: we are using `from __future__` import

**Comment by aarnphm** [comment]
> > for this feature I feel that we should use a requirements parser https://github.com/sarugaku/requirementslib or https://github.com/nexB/pip-requirements-parser
> 
> Why? Do we need to introduce an additional dependency for this fix/feature? Do we gain anything?

The reason for my proposal is that when user pass in their requirements file
```
--extra-index-url ...
```

Another contributor will also open a PR to support this feature in requirements.txt, and they might write their own parser. I think it is probably nicer if we use a parser and then that contributor can just use the same parser to add the feature.

**Comment by judahrand** [comment]
> The reason for my proposal is that when user pass in their requirements file
> 
> ```
> --extra-index-url ...
> ```
> 
> Another contributor will also open a PR to support this feature in requirements.txt, and they might write their own parser. I think it is probably nicer if we use a parser and then that contributor can just use the same parser to add the feature.

For your example, it would just work wouldn't it? `--extra-index-url` doesn't introduce an external filesystem dependency. Are there other examples which would currently not work?

**Comment by aarnphm** [comment]
> > for this feature I feel that we should use a requirements parser https://github.com/sarugaku/requirementslib or https://github.com/nexB/pip-requirements-parser
> 
> Why? Do we need to introduce an additional dependency for this fix/feature? Do we gain anything?

The reason for my proposal is that when user pass in their requirements file
```
--extra-index-url ...
```

Another contributor will also open a PR to support this feature in requirements.txt, and they might write their own parser. I think it is probably nicer if we use a parser, and that contributor can just use the same parser to add the feature.

**Comment by aarnphm** [comment]
`--extra-index-url` parse the URL right, which is not related to filesystem dependency.

File: tests/e2e/bento_server_http/bentofile.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply tests/e2e/bento_server_http/bentofile.yaml:3]: tbh this is more of a unit test cases, but we didn't write any unit tests for build config yet.
This e2e tests are more for testing HTTP functionality of bento server, not build requirements necessarily.

**Comment by judahrand** [comment]
> `--extra-index-url` parse the URL right, which is not related to filesystem dependency.

That's exactly my point. I'm not sure what other feature would be missing currently?

**Comment by aarnphm** [comment]
Right now, we only support a very minimal requirements.txt, and hence pretty much all of the advanced features from requirements.txt https://pip.pypa.io/en/stable/reference/requirements-file-format/#requirements-file-format are not yet supported

File: src/bentoml/_internal/bento/build_config.py (hunk 6 above)
Comment (sauyon) [reply src/bentoml/_internal/bento/build_config.py:561]: Yep, that makes sense to me. Can you catch that and throw a more user-friendly exception here that explains that we don't support it, then?

File: src/bentoml/_internal/utils/__init__.py [hunk 7]
```diff
@@ -303,6 +304,24 @@ def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str:
...
+def find_nested_requirements_txts(
+    base_requirements_txt: Path,
+) -> tuple[Path, set[Path]]:
+    """Recursively search the provided requirements.txt file to find nested
+    requirements files specified with `-r`.
+    """
+    sub_requirements_txt_files: set[Path] = set()
+    with base_requirements_txt.open("r") as f:
+        for line in f.readlines():
+            if line.startswith("-r"):
```
Comment (sauyon) [inline src/bentoml/_internal/utils/__init__.py:316]: Maybe something like this so we support both `--requirement` and whitespace.

```suggestion
            if re.match(r"^\s*(-r|--requirement)", line):
```

File: src/bentoml/_internal/utils/__init__.py [hunk 8]
```diff
@@ -303,6 +304,24 @@ def resolve_user_filepath(filepath: str, ctx: t.Optional[str]) -> str:
...
+                base, nested = find_nested_requirements_txts(
+                    base_requirements_txt.parent / line.split("-r")[1].strip(),
```
Comment (sauyon) [inline src/bentoml/_internal/utils/__init__.py:318]: In case there's `-r` inside a filename. Maybe you'll need another branch for `--requirement` too.

```suggestion
                    base_requirements_txt.parent / line.split("-r", 1)[1].strip(),
```

**Review by sauyon — COMMENTED** [review]
While I also think eventually we may want to switch to using a library for this just using a regex for now is probably fine, since it's not too much overhead and we're only supporting this one feature. Might be worth double checking their implementation to ensure that we aren't missing some other syntax, though.

File: tests/e2e/bento_server_http/bentofile.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (judahrand) [reply tests/e2e/bento_server_http/bentofile.yaml:3]: Yeah, I agree that it is something that should really be included in unit testing rather than the e2e tests. However, I don't see the utility in introducing an entirely new e2e test (which will increase the 'test time') to cover it. Really at some point the build requirements should be unit tested and then this e2e test could go back to how it was -- this isn't something I think should be tackled in this PR though.

I can if you really want but it doesn't feel necessary.

File: src/bentoml/_internal/bento/build_config.py (hunk 6 above)
Comment (judahrand) [reply src/bentoml/_internal/bento/build_config.py]: Now that I've switched to using `pip-requirements-parser` as @aarnphm suggested this should no longer be a limitation as the parser is reading all nested files and essentially concating them all into a single file (with a small modification to make sure that we remove other file references).

I think the only thing which might not work not is `--find-links` with a path to an HTML file. However, that seems like a super niche edge case.

**Comment by judahrand** [comment]
Now that I've switched to using pip-requirements-parser as @aarnphm suggested this should no longer be a limitation as the parser is reading all nested files and essentially concating them all into a single file (with a small modification to make sure that we remove other file references).

I think the only thing which might not work not is --find-links with a path to an HTML file. However, that seems like a super niche edge case.

**Comment by judahrand** [comment]
I don't think that the current test failures are actually related to my changes. They seem to be a combination of CodeCov upload failures and tests which are also failing in `main`. Are these able to be fixed?

File: tests/e2e/bento_server_http/bentofile.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply tests/e2e/bento_server_http/bentofile.yaml:3]: Me and @sauyon discussed that this change is ok for this PR. We will add unit tests for build config specifically to test out these use case.

**Comment by aarnphm** [comment]
> I don't think that the current test failures are actually related to my changes. They seem to be a combination of CodeCov upload failures and tests which are also failing in `main`. Are these able to be fixed?

I have a PR to fix the tests, for that one to get merged first, and then I will rebase and update your changes into the main branch.

**Comment by aarnphm** [comment]
gg

> Now that I've switched to using `pip-requirements-parser` as @aarnphm suggested the only thing in a `requirements.txt` file which might not work not is `--find-links` with a path to an HTML file. However, that seems like a super niche edge case.

Thanks for the reminder. I will follow up with a PR for documentation about this behaviour.

**Review by aarnphm — APPROVED** [review]
